docs/design/operator.md

```diff
@@ -0,0 +1,387 @@
+  second instance of a driver of a similar type, it should be deployed in a
+  different namespace.
+- Each CSI driver must have a unique name across all namespaces.
+
+```yaml
+---
+kind: CephCSIDriver
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: "<prefix>.<driver_type>.csi.ceph.com"
+  namespace: <operator-namespace>
+spec:
+  fsGroupPolicy: File
+  encryption:
+    configMapRef:
+      name: encryption-config-map-name
+  plugin:
+    priorityClassName: system-node-critical
+    updateStrategy:
+      type: RollingUpdate
+      rollingUpdate:
+        maxUnavailable: 1
+    labels:
+      app: cephfs-plugin
+    annotations:
+      k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+  provisioner:
+    labels:
+      app: ceph-fs-provisioner
+    annotations:
+      k8s.v1.cni.cncf.io/networks: macvlan-conf-1
+    provisionerReplicas: 2
+    leaderElection:
+      leaseDuration: 100
+      renewDeadline: 100
+      retryPeriod: 10
+    attachRequired: true
+  liveness:
+    metricsPort: 8000
+  deployCSIAddons: false
+  kernelClient: true
+  kernelMountOptions: ms_mode=secure
+status:
+  phase: Failed
+  reason: csi driver with same name already exists in the cluster
+```
+
+### CephCSICephConfig CRD
+
+Stores connection and configuration details for a single Ceph cluster and
+provide the information to be used by multiple CSI drivers.
+
+```yaml
+---
+kind: CephCSICephCluster
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: ceph-cluster-1
+  namespace: <operator-namespace>
+spec:
+  monitors:
+  - 10.98.44.171:6789
+  - 10.98.44.172:6789
+  - 10.98.44.173:6789
+  ReadAffinity:
+    crushLocationLabels:
+    - kubernetes.io/hostname
+    - topology.kubernetes.io/region
+    - topology.kubernetes.io/zone
+  rbdMirrorDaemonCount: 2
+  cephConfig:
+    global:
+      auth_cluster_required: none
+      auth_service_required: none
+      auth_client_required: none
+      rbd_validate_pool: false
+status: {}
+```
+
+### CephCSIConfig CRD
+
+Contains details about CephFS, RBD, and NFS configuration to be used when
+communicating with Ceph. Include a reference to a CephCSICephCluster holding
+the connection information for the target Ceph cluster.
+
+```yaml
+---
+kind: CephCSIConfig
+apiVersion: csi.ceph.io/v1alpha1
+metadata:
+  name: storage
+  namespace: <operator-namespace>
+spec:
+  cephClusterRef:
+    name: ceph-cluster-1
+  cephFS:
+    subvolumeGroup: csi
+    kernelMountOptions: readdir_max_bytes=1048576,norbytes
+    fuseMountOptions: debug
+  rbd:
+    radosNamespace: rados-test
+status:
+  phase: Succeeded
+  reason: successfully linked to CephClusterRef
+```
+
+### CephCSIConfigMapping
+
+The CephCSIConfigMapping CR contains a mapping between local and remote Ceph
+cluster configurations. The information in this CR helps Ceph-CSI identify
+peered blockpools. This can help with the correct management and consumption of
+volumes in backup and DR scenarios
+
+```yaml
+---
+kind: CephCSIConfigMapping
```

Comment:
I think I commented on this previously, but can't find it now or if there was a response... "Config" in the CRD name is redundant IMO. How about a name without "Config"? For example, CephCSIPeerMapping, CephCSIPoolMapping, CephCSIClusterMapping, or even just CephCSIMapping
```suggestion
kind: CephCSIPeerMapping
```

Reply:
Here config is a subject, not an adjunctive. So it makes sense to keep it.

Reply:
My point is that everything in every CRD is config, why does it need to be part of the CRD name?

Reply:
I can see Travis' point here. It's obvious that it's a subject, but it seems redundant in the sense that all k8s resources contain config (or spec, as k8s would call it). Is there a more clear name for this/these?

CephCSIPeerMapping does seem to be more descriptive and explanatory to me, based on the description:

> The CephCSIConfigMapping CR contains a mapping between local and remote Ceph
cluster configurations. The information in this CR helps Ceph-CSI identify
peered blockpools. This can help with the correct management and consumption of
volumes in backup and DR scenarios

IMO, even "mapping" is a bit confusing here. I feel that "mapping" may be too unclear, or just redundant.

IMO, CephCSIPeer is a sufficient name to indicate that
1. This configures Ceph-CSI
2. This configures Ceph-CSI with "peer" information

Reply:
IMO `CephCSIConfigMapping` is the right name for it because this collates the relationship of two things here only is the ceph pool mapping and also the csi configuration Name `cephCSIConfigName` , and its named a `CephCSIConfigMapping` because its a Mapping to the `CephCSIConfigMapping`, 


The usecase of this is that there are PVC exists with pool A but for some reason pool A was recover but it got a new pool ID for that cases, the user can still creating this CR to allow cephcsi to work with existing PVC's (this is totally low level on how cephcsi work and how we can recover few things), Currently we don't have support for this in Rook but we have support for this in cephcsi. This CR is really targeting to cover multiple users (cases).

>CephCSIPeerMapping does seem to be more descriptive and explanatory to me, based on the description:

The CephCSIConfigMapping CR contains a mapping between local and remote Ceph
cluster configurations. The information in this CR helps Ceph-CSI identify
peered blockpools. This can help with the correct management and consumption of
volumes in backup and DR scenarios

This was one more details that missing in the description but it was present in the CR field. i added more details to it i hope now this makes sense.

Reply:
The information inside this CR maps between entities of type `CephCSIConfig`. This is why it is natural to call it `CephCSIConfig + Mapping = CephCSIConfigMaping` 

The reason we call the original entity `CephCSIConfig` is because it represents a configuration entry for CSI, we cannot just drop the config part. If we do so it becomes `CephCSI` which does not make sense. 

Regarding the `CephCSIOperatorConfig` name: A config map of a similar name is a common practice for operators (`<operator-name>-operator-config`), and is even provided by default. We just elevated the concept into a CR

Reply:
I still am of the opinion that all of these would be better named without "Config", but let's take it as a separate discussion point and not block the design PR. Naming always takes time to settle. How about we open a separate issue for the discussion?